packages/contracts-bedrock/src/L2/CrossL2Inbox.sol

```diff
@@ -0,0 +1,139 @@
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.blocknumber")) - 1)
+    bytes32 public constant BLOCKNUMBER_SLOT = 0x5a1da0738b7fdc60047c07bb519beb02aa32a8619de57e6258da1f1c2e020ccc;
+
+    /// @notice Transient storage slot that the logIndex for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.logindex")) - 1)
+    bytes32 public constant LOG_INDEX_SLOT = 0xab8acc221aecea88a685fabca5b88bf3823b05f335b7b9f721ca7fe3ffb2c30d;
+
+    /// @notice Transient storage slot that the timestamp for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.timestamp")) - 1)
+    bytes32 public constant TIMESTAMP_SLOT = 0x2e148a404a50bb94820b576997fd6450117132387be615e460fa8c5e11777e02;
+
+    /// @notice Transient storage slot that the chainId for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.chainid")) - 1)
+    bytes32 public constant CHAINID_SLOT = 0x6e0446e8b5098b8c8193f964f1b567ec3a2bdaeba33d36acb85c1f1d3f92d313;
+
+    /// @custom:semver 1.0.0
+    string public constant version = "1.0.0";
+
+    /// @notice Returns the origin address of the Identifier.
+    /// @return _origin The origin address of the Identifier.
+    function origin() external view returns (address _origin) {
+        assembly {
+            _origin := tload(ORIGIN_SLOT)
+        }
+    }
+
+    /// @notice Returns the block number of the Identifier.
+    /// @return _blocknumber The block number of the Identifier.
+    function blocknumber() external view returns (uint256 _blocknumber) {
+        assembly {
+            _blocknumber := tload(BLOCKNUMBER_SLOT)
+        }
+    }
+
+    /// @notice Returns the log index of the Identifier.
+    /// @return _logIndex The log index of the Identifier.
+    function logIndex() external view returns (uint256 _logIndex) {
+        assembly {
+            _logIndex := tload(LOG_INDEX_SLOT)
+        }
+    }
+
+    /// @notice Returns the timestamp of the Identifier.
+    /// @return _timestamp The timestamp of the Identifier.
+    function timestamp() external view returns (uint256 _timestamp) {
+        assembly {
+            _timestamp := tload(TIMESTAMP_SLOT)
+        }
+    }
+
+    /// @notice Returns the chain ID of the Identifier.
+    /// @return _chainId The chain ID of the Identifier.
+    function chainId() external view returns (uint256 _chainId) {
+        assembly {
+            _chainId := tload(CHAINID_SLOT)
+        }
+    }
+
+    /// @notice Executes a cross chain message on the destination chain.
+    /// @param _id An Identifier pointing to the initiating message.
+    /// @param _target Account that is called with _msg.
+    /// @param _msg The message payload, matching the initiating message.
+    function executeMessage(Identifier calldata _id, address _target, bytes memory _msg) public payable {
```

Comment:
The ordering of the calldata doesn't follow the spec

Reply:
I included `calldata` first so that I didn't have to pad `calldataload` when loading `_id` in assembly (need this for storing in transient storage). if we add the padding, we risk new vars introduced in the future (and other things) from breaking the function. what do you think?

Reply:
If there are new variables added in the future, the function selector will be different. This ABI will never change, we will only expand on the ABI of this contract. Any reduction in ABI will need to go through a deprecation process and take a long time

Reply:
It is very important that the calldata definition lines up here and in the Go code that parses the calldata. I still have a preference for the way that the spec defines it but I am open to modifying it if we can show that its far easier to do it a different way. There is both a solidity and go implementation of the calldata parsing, so the alteration would impact both